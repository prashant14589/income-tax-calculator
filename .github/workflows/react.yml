name: React CI

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - name: Checkout
      uses: actions/checkout@v2
    - name: Install dependencies
      run: npm install
    - name: Run jest
      run: ./node_modules/.bin/jest src --passWithNoTests
    - name: Cypress run
      uses: cypress-io/github-action@v2
      if: always()
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      with:
        build: npm run build --if-present
        start: npm start
        wait-on: 'http://localhost:3000'
