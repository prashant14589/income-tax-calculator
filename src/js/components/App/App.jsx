import React, { useCallback, useEffect, useState } from 'react';
import Select from "../SelectOption/Select";
import { TextInput, Wrapper } from '../Input/TextInput';
import calculateTax from "../../tax-calculator-helper";

export const App = () => {
    const yearOptions = [
        { value: "", label: "Please enter year" },
        { value: "2018", label: "2018" },
        { value: "2019", label: "2019" },
        { value: "2020", label: "2020" },
    ];
    const [year, setYear] = useState(0);
    const [age, setAge] = useState(0);
    const [income, setIncome] = useState(0);
    const [investment, setInvestment] = useState(0);
    const [finalTaxPayble, setFinalTaxPayble] = useState(0);
    const [displayTaxCalculator, setDisplayTaxCalculator] = useState(true);
    // TODO - need to cehck if useCallback is better than useEfect here?
    const finalTaxCalculation = useEffect(
        () => {
            setFinalTaxPayble(calculateTax(year, age, income, investment));
        },
        [year, age, income, investment],
    );
    return (
        <>
            {/* <Login ></Login> */}
            <>
                <section>
                    <h1>Tax Calculator</h1>
                    <label htmlFor="year">Year:
    <Select
                            options={yearOptions}
                            value={year}
                            handleChange={(e) => setYear(e.target.value)}
                        />
                    </label>

                    <TextInput id="age" value={age} onChange={(e) => setAge(e.target.value)}>Age:</TextInput>

                    <TextInput id="income" value={income} onChange={(e) => setIncome(e.target.value)}>Income:</TextInput>

                    <TextInput id="investment" value={investment} onChange={(e) => setInvestment(e.target.value)}>Investment:</TextInput>

                </section>
                <TextInput id="tax" value={finalTaxPayble} readOnly={true}>Final Tax calculated:</TextInput>
            </>

        </>
    )
}
