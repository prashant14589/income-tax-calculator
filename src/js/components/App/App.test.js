import React from 'react';
import {fireEvent, render, screen } from '@testing-library/react';
import { App } from './App';

describe('<App />', () => {
  beforeEach(() => {
    Object.defineProperty(window, "localStorage", {
      value: {
        getItem: jest.fn(() => null),
        setItem: jest.fn(() => null)
      },
      writable: true
    });
    window.localStorage.setItem("user","A");
  });
  it('should render successfully', () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });  
  beforeEach( () => {
    render(<App />);
  });
  it('should render select options successfully', () => {
    
    expect(screen.getByText('Year:')).toBeInTheDocument();
  });
  it('should render input fields successfully', () => {
    
    expect(screen.getByText('Age:')).toBeInTheDocument();
  });
  it('should render input fields successfully', () => {
    
    expect(screen.getByText('Income:')).toBeInTheDocument();
  });
  it('should render input fields successfully', () => {
    
    expect(screen.getByText('Investment:')).toBeInTheDocument();
  });
  it('should render input fields successfully', () => {
   
    expect(screen.getByText('Final Tax calculated:')).toBeInTheDocument();
  });
  
});
describe("expect valid output for calculated tax", () => {
  render(<App />);
  const age = screen.getByRole("textbox", { name: 'Age:'});
  const income = screen.getByRole("textbox", { name: 'Income:'});
  const investment = screen.getByRole("textbox", { name: 'Investment:'});
  const tax = screen.getByRole("textbox", { name: 'Final Tax calculated:'});
  fireEvent.change(screen.getByTestId("select"),{
    target: { value: '2020' },
  });
  fireEvent.change(age, {
    target: { value: '65' },
  });
  
  fireEvent.change(income, {
    target: { value: '3800000' },
  });
  fireEvent.change(investment, {
    target: { value: '100000' },
  });
  expect(tax.value).toBe("830812.5")
})
