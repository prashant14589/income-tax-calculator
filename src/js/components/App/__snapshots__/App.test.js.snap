// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render successfully 1`] = `
.c0 {
  width: 30%;
  height: 35px;
  background: papayawhip;
  color: gray;
  padding-left: 5px;
  font-size: 14px;
  border: none;
  margin-left: 10px;
}

.c0 option {
  color: palevioletred;
  background: papayawhip;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  white-space: pre;
  min-height: 20px;
  padding: 0px 2px 1px;
}

.c1 {
  padding: 0.5em;
  display: inline-block;
  width: 30%;
  margin: 0.5em;
  color: palevioletred;
  background: papayawhip;
  border: none;
  border-radius: 3px;
}

<div>
  <section>
    <h1>
      Tax Calculator
    </h1>
    <label
      for="year"
    >
      Year:
      <select
        class="c0"
        data-testid="select"
      >
        <option
          value=""
        >
          Please enter year
        </option>
        <option
          value="2018"
        >
          2018
        </option>
        <option
          value="2019"
        >
          2019
        </option>
        <option
          value="2020"
        >
          2020
        </option>
      </select>
    </label>
    <div>
      <label
        for="age"
      >
        Age:
      </label>
      <input
        class="c1"
        id="age"
        type="text"
        value="0"
      />
    </div>
    <div>
      <label
        for="income"
      >
        Income:
      </label>
      <input
        class="c1"
        id="income"
        type="text"
        value="0"
      />
    </div>
    <div>
      <label
        for="investment"
      >
        Investment:
      </label>
      <input
        class="c1"
        id="investment"
        type="text"
        value="0"
      />
    </div>
  </section>
  <div>
    <label
      for="tax"
    >
      Final Tax calculated:
    </label>
    <input
      class="c1"
      id="tax"
      readonly=""
      type="text"
      value="0"
    />
  </div>
</div>
`;
